File: Coin Change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        df = [0]+[float('inf')]*amount 
        for i in range(1, amount+1):
            for coin in coins:
                if i-coin >= 0:
                    df[i] = min(df[i],df[i-coin]+1)
        if df[-1] == float('inf'):
            return -1
        return df[-1]

File: Coin Change.js
var coinChange = function(coins, amount) {
    const dfs = (target, memo={}) => {
        if(target in memo) return memo[target]
        if(target<0) return -1
        else if (target===0) return 0
        let smallestCount = Infinity
        for(let i=0; i<coins.length; i++) {
            let count = dfs(target-coins[i], memo)
            if (count>=0 && count<smallestCount) {
                smallestCount = count + 1
            }
        }
        memo[target] = smallestCount === Infinity ? -1 : smallestCount
        return memo[target]
    }
    return dfs(amount)
};

console.log(coinChange([1,2,5], 11))